### ISSUE TITLE
ChucK 1.4 Crashing on Windows

### ISSUE THREAD

**Issue by paranoid2droid** [ccrma/chuck#118]
Environment:
Chuck 1.4
Windows 7 Professional

I was taking the physics DSP course on Kadenze.com. It's fascinating. However, when I run the sample code in Coursework1, ChucK Crashed, either in CMD or in miniAudicle.

On the forum of the course there are also people ran into the same thing on Windows 10.

> Viκtor Markos Karamanis Posted By Viκtor Markos Karamanis
> Mar 31, 2018 @ 3:28amCST
> Hello,
> 
> I try to open the pong.bat but chuck is crashing.
> 
> i have the latest (1.4) 
> 
> i think is something wrong with the GL as i try to open the PongViewGLWin32.exe and 64 also and it outputs that freeglut.dll | glut64.dll is missing 
> 
> Searched for that and putted freeglut.dll in the same folder, PongViewGLWin32.exe opens up but pong still not opening
> 
> any ideas?

Anyway, it works if when ChucK rolls back to 1.3.5.2.

The problem code:
```
// PongPhysics1.ck   by Perry R. Cook  2015
// Simple physics of bouncing ball in a 
//    square  box.  One peg in center.
//  Collisions trigger sounds.  

// Simple Ball object with physics variables
class Ball {
    0.06 => float r;   // radius
    0.5 => float x;    // x position
    0.5 => float y;    // y position
    0.02 => float dx;  // velocity in x
    0.01 => float dy;  // velocity in y
    0.0 => float ax;   // acceleration in x 
    0.0 => float ay;   // acceleration in y
    0.995 => float damp; // 1.0 means no loss over time

    fun void doPhysics()  {
        ax +=> dx;   // dx += ax; // in C/Java
        ay +=> dy;   // dy += ay; // in C/Java
        (damp *=> dx) +=> x;      // x += (dx *= damp);  // in C/Java
        (damp *=> dy) +=> y;      // y += (dy *= damp);  // in C/Java
    }
}

// Define your sound making objects here
SndBuf ballSound => Pan2 pan => dac;
SndBuf pegSound => dac;
"special:dope" => ballSound.read;
ballSound.samples() => ballSound.pos;
me.dir()+"Synthesize.wav" => pegSound.read;
pegSound.samples() => pegSound.pos;

Ball ball; // moving ball
Ball peg;  // rigid filled peg
0.0 => peg.x => peg.y => peg.dx => peg.dy;
drawPeg(0,peg);  // only have to do this once

now + 20::second => time then;  // run for 10 seconds

while (now < then)   {
    ball.doPhysics();
    drawBall(1,ball);
    if (checkLeftRightWalls(ball)) {
        ball.x => pan.pan;  // pan is set by position
        soundBall(ball.dx);
    }
    if (checkTopBottomWalls(ball)) {
        ball.x => pan.pan;
        soundBall(ball.dy);
        -ballSound.rate() => ballSound.rate;  // play backward!!
        ballSound.samples() => ballSound.pos;
    }   
    if (checkBallPegCollide(ball,peg)) {
        soundPeg(ball.dx+ball.dy);
    }
    33*ms => now;  // 30 frames (roughly) per second
}

chout <= "Exit\n";   // send exit signal to GL Viewer
chout.flush();

fun void drawBall(int which,Ball bl)  {
    chout <= "b"+which+" "+bl.x+" "+bl.y+" "+bl.r+"\n";   // draw a ball
    chout.flush();                  // make sure things don't linger in buffers
}

fun void drawPeg(int which, Ball pg)  {
    drawBall(which,pg);        // draw a ball, but
    chout <= "f"+which+" 1\n"; // then fill it in
    chout.flush();             // make sure things don't linger in buffers
}

fun int checkLeftRightWalls(Ball bl)  {
    if (bl.x < -(1.0-bl.r)) {  // check left limit in x
        3*bl.dx -=> bl.x;  // bounce off
        -1 *=> bl.dx;   // with velocity reflection
        Math.random2f(-0.001,0.001) => bl.ax;  // new random x force
        return 1;
    }
    else if (bl.x > (1.0-bl.r)) {  // check right limit in x
        3*bl.dx -=> bl.x;  // bounce off
        -1 *=> bl.dx;   // with velocity reflection
        Math.random2f(-0.001,0.001) => bl.ax;  // new random x force
        return 1;
    }
    else return 0;
}

fun int checkTopBottomWalls(Ball bl)  {
    if (bl.y > (1.0-bl.r)) {  // check top limit in y
        3*bl.dy -=> bl.y;  // bounce off
        -1 *=> bl.dy;   // with velocity reflection
        Math.random2f(-0.001,0.001) => bl.ay;  // new random y force
        return 1;
    }
    else if (bl.y < -(1.0-bl.r)) {  // check top limit in y
        3*bl.dy -=> bl.y;  // bounce off
        -1 *=> bl.dy;   // with velocity reflection
        Math.random2f(-0.001,0.001) => bl.ay;  // new random y force
        return 1;
    }
    else return 0;
}

fun float distance(Ball bl, Ball bl2)  {
    bl2.x - bl.x => float delx;
    bl2.y - bl.y => float dely;
    return Math.sqrt(delx*delx + dely*dely);
}

fun int checkBallPegCollide(Ball bl, Ball pg)  {
    if (distance(bl,pg) < bl.r+pg.r)  {  // collision!!
        3*bl.dy -=> bl.y;  // bounce off
        -0.5 *=> bl.dy;   // with velocity reflection and great loss
        3*bl.dx -=> bl.x;  // bounce off
        -0.5 *=> bl.dx;   // with velocity reflection and great loss

        // the peg has magical powers!!!
        Math.random2f(-0.001,0.001) => bl.ax; // new random forces
        Math.random2f(-0.001,0.001) => bl.ay; //  after collision
        return 1;
    }
    else return 0;
}

fun void soundPeg(float vel)  {
    0.2+vel => pegSound.gain;
    0.8 + vel*2.0 => pegSound.rate;
    0 => pegSound.pos;
}

fun void soundBall(float vel)  {
    0.2+vel => ballSound.gain;
    0.8 + vel*2.0 => ballSound.rate;
    0 => ballSound.pos;
}
```

**Comment by BSWheeler42** [comment]
This has been solved.
Add 2 lines to chuck_oo.cpp, and recompile.
In Chuck_IO_Chout::Chuck_IO_Chout( Chuck_Carrier * carrier ) {...
at 3112 add m_callback = NULL;
In Chuck_IO_Cherr::Chuck_IO_Cherr( Chuck_Carrier * carrier ) {...
at 3204 add m_callback = NULL;
In both classes, m_callback is tested against NULL later, without first being initialized.
I have recompiled chuck 1.4.0.1 and tested on my PC.

**Comment by BSWheeler42** [comment]
According to Ge Wang,
This has been fixed and merge into main (as part of PR #157) and will be included in 1.4.0.2.
